In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r"C:\Users\mw3595\Documents\data vis project\311_Service_Requests_2024_2025.csv",on_bad_lines="skip")

C:\Users\mw3595\AppData\Local\Temp\ipykernel_13140\1061516281.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r"C:\Users\mw3595\Documents\data vis project\311_Service_Requests_2024_2025.csv",on_bad_lines="skip")


https://data.cityofnewyork.us/Social-Services/311-Service-Requests-from-2020-to-Present/erm2-nwe9/explore/query/SELECT%0A%20%20%60unique_key%60%2C%0A%20%20%60created_date%60%2C%0A%20%20%60closed_date%60%2C%0A%20%20%60agency%60%2C%0A%20%20%60agency_name%60%2C%0A%20%20%60complaint_type%60%2C%0A%20%20%60descriptor%60%2C%0A%20%20%60location_type%60%2C%0A%20%20%60incident_zip%60%2C%0A%20%20%60incident_address%60%2C%0A%20%20%60street_name%60%2C%0A%20%20%60cross_street_1%60%2C%0A%20%20%60cross_street_2%60%2C%0A%20%20%60intersection_street_1%60%2C%0A%20%20%60intersection_street_2%60%2C%0A%20%20%60address_type%60%2C%0A%20%20%60city%60%2C%0A%20%20%60landmark%60%2C%0A%20%20%60facility_type%60%2C%0A%20%20%60status%60%2C%0A%20%20%60due_date%60%2C%0A%20%20%60resolution_description%60%2C%0A%20%20%60resolution_action_updated_date%60%2C%0A%20%20%60community_board%60%2C%0A%20%20%60bbl%60%2C%0A%20%20%60borough%60%2C%0A%20%20%60x_coordinate_state_plane%60%2C%0A%20%20%60y_coordinate_state_plane%60%2C%0A%20%20%60open_data_channel_type%60%2C%0A%20%20%60park_borough%60%2C%0A%20%20%60latitude%60%2C%0A%20%20%60longitude%60%2C%0A%20%20%60location%60%0AWHERE%0A%20%20%60created_date%60%0A%20%20%20%20BETWEEN%20%222024-01-01T16%3A23%3A16%22%20%3A%3A%20floating_timestamp%0A%20%20%20%20AND%20%222025-12-31T16%3A23%3A16%22%20%3A%3A%20floating_timestamp%0A%20%20AND%20caseless_ne%28%60agency%60%2C%20%22HPD%22%29%0AORDER%20BY%20%60created_date%60%20DESC%20NULL%20FIRST/page/filter


#### The downloaded data contains only 911 complaints for 2024 and 2025. File excludes HPD (New York City Department of Housing Preservation and Development) because it is usually concerned only with building conditions and is not relevant to our project conceptually.


**Data also does not include those columns:**
    "vehicle_type",
    "taxi_company_borough",
    "taxi_pick_up_location",
    "bridge_highway_name",
    "bridge_highway_direction",
    "road_ramp",
    "bridge_highway_segment",
    "due_date", 
    "resolution_description",
    "resolution_action_updated_date",
    "landmark",
    "facility_type" 

In [3]:
df.head()

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Problem (formerly Complaint Type),Problem Detail (formerly Descriptor),Location Type,Incident Zip,Incident Address,...,Community Board,BBL,Borough,X Coordinate (State Plane),Y Coordinate (State Plane),Open Data Channel Type,Park Borough,Latitude,Longitude,Location
0,67347796,12/31/2025 04:23:00 PM,01/01/2026 01:00:00 AM,DEP,Department of Environmental Protection,Water System,Dirty Water (WE),NaN,10031.0,626 RIVERSIDE DRIVE,...,09 MANHATTAN,1.020870e+09,MANHATTAN,"996,862","239,357",ONLINE,MANHATTAN,40.823646,-73.954431,POINT (-73.954430909076 40.82364624134)
1,67346303,12/31/2025 04:22:33 PM,01/02/2026 01:35:00 PM,DOT,Department of Transportation,Street Condition,Pothole,NaN,10461.0,INDUSTRIAL STREET,...,11 BRONX,NaN,BRONX,NaN,NaN,UNKNOWN,BRONX,NaN,NaN,NaN
2,67347638,12/31/2025 04:22:31 PM,01/02/2026 08:31:35 AM,DOT,Department of Transportation,Street Sign - Missing,Other/Unknown,Street,10314.0,4113 VICTORY BOULEVARD,...,02 STATEN ISLAND,5.027670e+09,STATEN ISLAND,"931,214","154,849",MOBILE,STATEN ISLAND,40.591542,-74.190961,POINT (-74.190961278787 40.591542246598)
3,67346505,12/31/2025 04:22:09 PM,01/07/2026 08:54:28 PM,DSNY,Department of Sanitation,Dirty Condition,Trash,Street,10026.0,ST NICHOLAS AVENUE,...,10 MANHATTAN,NaN,MANHATTAN,"997,313","231,960",PHONE,MANHATTAN,40.803343,-73.952816,POINT (-73.952815773413 40.803342887744)
4,67351565,12/31/2025 04:21:58 PM,12/31/2025 04:37:20 PM,NYPD,New York City Police Department,Noise - Vehicle,Car/Truck Horn,Street/Sidewalk,11366.0,78-15 PARSONS BOULEVARD,...,08 QUEENS,4.068290e+09,QUEENS,"1,037,000","202,363",ONLINE,QUEENS,40.721959,-73.809697,POINT (-73.809696824262 40.721959131993)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5597361 entries, 0 to 5597360
Data columns (total 33 columns):
 #   Column                                Dtype  
---  ------                                -----  
 0   Unique Key                            int64  
 1   Created Date                          object 
 2   Closed Date                           object 
 3   Agency                                object 
 4   Agency Name                           object 
 5   Problem (formerly Complaint Type)     object 
 6   Problem Detail (formerly Descriptor)  object 
 7   Location Type                         object 
 8   Incident Zip                          object 
 9   Incident Address                      object 
 10  Street Name                           object 
 11  Cross Street 1                        object 
 12  Cross Street 2                        object 
 13  Intersection Street 1                 object 
 14  Intersection Street 2                 object 
 15  Address Type   

In [5]:
df["Created Date"] = pd.to_datetime(
    df["Created Date"],
    format="%m/%d/%Y %I:%M:%S %p",
    errors="coerce"
)

df["Closed Date"] = pd.to_datetime(
    df["Closed Date"],
    format="%m/%d/%Y %I:%M:%S %p",
    errors="coerce"
)

In [6]:
df["Complaint Year"] = df["Created Date"].dt.year

In [7]:
df["Complaint Month"] = df["Created Date"].dt.month

In [8]:
print(df["Complaint Year"].unique())
df["Complaint Month"].unique()

[2025 2024]


array([12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

In [9]:
df["Month Name"] = pd.to_datetime(df["Complaint Month"], format="%m").dt.month_name()

In [10]:
df["Month Name"].unique()

array(['December', 'November', 'October', 'September', 'August', 'July',
       'June', 'May', 'April', 'March', 'February', 'January'],
      dtype=object)

In [14]:
df["Agency"].unique()

array(['DEP', 'DOT', 'DSNY', 'NYPD', 'DOB', 'DOHMH', 'OOS', 'TLC', 'EDC',
       'DCWP', 'DPR', 'DHS', 'DOE', 'OTI', '3-1-1'], dtype=object)

In [11]:
#df = df[df["Agency"] != "HPD"]
#df["Agency"].unique()

In [12]:
#df["Problem (formerly Complaint Type)"].unique()

In [13]:
#for the ease of use - we will rename the columns
df = df.rename(columns={
    "Problem (formerly Complaint Type)": "Complaint",
    "Problem Detail (formerly Descriptor)": "Complaint Detail"
})

In [14]:
df = df.rename(columns={
    'X Coordinate (State Plane)': "X Coordinate",
    'Y Coordinate (State Plane)': "Y Coordinate"
})

In [15]:
grouping_complaints = (
    df.groupby(["Agency", "Complaint","Complaint Detail"])
      .size()
      .reset_index(name="count")
)

In [16]:
grouping_complaints

,Agency,Complaint,Complaint Detail,count
0,3-1-1,Noise - Commercial,Loud Music/Party,1
1,DCWP,Consumer Complaint,Airline,8
2,DCWP,Consumer Complaint,Arcade,9
3,DCWP,Consumer Complaint,Art Supply Store,11
4,DCWP,Consumer Complaint,Auction House or Auctioneer,18
...,...,...,...,...
1142,TLC,Taxi Licensee Complaint,Owner/Broker/Agent Complaint,4
1143,TLC,Taxi Licensee Complaint,Vehicle Report,1
1144,TLC,Taxi Report,Driver Report - Non Passenger,314
1145,TLC,Taxi Report,Driver Report - Passenger,1103


In [17]:
#if you want your own table :
# grouping_complaints.to_excel("complaints_by_agency_problem.xlsx", index=False)

In [18]:
df.isnull().sum()

Unique Key                              0
Created Date                            0
Closed Date                        136499
Agency                                  0
Agency Name                             0
Complaint                               0
Complaint Detail                   196265
Location Type                      845736
Incident Zip                        64029
Incident Address                   242567
Street Name                        242798
Cross Street 1                     501565
Cross Street 2                     500465
Intersection Street 1              782735
Intersection Street 2              779527
Address Type                        28434
City                               317112
Landmark                          1267524
Facility Type                     5571498
Status                                  0
Due Date                          5561175
Resolution Description             138109
Resolution Action Updated Date      56759
Community Board                   

We have 5,597,360 rows and for:
- 102,672 we do not have Location
- 102,639 we do not have X Coordinate
- 102,458 we do not have Y Coordinate

In [19]:
df[(df["Location"].notna()) & (df["X Coordinate"].isna())]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [20]:
rows_with_location_missing_x = df[
    df["Location"].notna() & df["X Coordinate"].isna()
]

print("Number of rows:", len(rows_with_location_missing_x))  # should print 33
rows_with_location_missing_x.head()

Number of rows: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [21]:
df[(df["Location"].notna()) & (df["Y Coordinate"].isna())]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [22]:
df[["Location","X Coordinate"]].isna().value_counts()

Location  X Coordinate
False     False           5494689
True      True             102639
          False                33
Name: count, dtype: int64

5,494,689 rows → both Location and X Coordinate (State Plane) exist ✅

102,639 rows → Location is missing, X Coordinate is missing ❌

33 rows → X Coordinate exists but Location is missing ⚠

In [23]:
rows_with_location_missing_x = df[
    df["X Coordinate"].notna() & df["Location"].isna()
]

len(rows_with_location_missing_x)  # should print 33
rows_with_location_missing_x

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
83653,67224277,2025-12-17 14:44:57,2025-12-18 14:15:05,DOT,Department of Transportation,Street Condition,Blocked - Construction,Street,11201.0,CLARK STREET,...,"985,392",NaN,PHONE,BROOKLYN,NaN,NaN,NaN,2025,12,December
176201,67082296,2025-12-06 17:31:33,2025-12-09 08:44:14,DOT,Department of Transportation,Broken Parking Meter,Out of Order,Street,10075.0,EAST 78 STREET,...,"994,680",NaN,ONLINE,MANHATTAN,NaN,NaN,NaN,2025,12,December
298168,66901503,2025-11-20 16:20:18,NaT,DOE,Department of Education,School Maintenance,Other School Condition,School,NaN,62-10,...,"1,025,093",NaN,ONLINE,QUEENS,NaN,NaN,NaN,2025,11,November
346429,66829282,2025-11-13 19:02:33,2025-11-17 09:25:16,DOT,Department of Transportation,Street Condition,"Rough, Pitted or Cracked Roads",Street,10452.0,RIVER AVENUE,...,"1,006,269",NaN,PHONE,BRONX,NaN,NaN,NaN,2025,11,November
396605,66759483,2025-11-07 06:53:45,2025-11-10 12:57:50,DOT,Department of Transportation,Street Condition,Cave-in,Street,11208.0,RIDGEWOOD AVENUE,...,"1,017,810",NaN,UNKNOWN,BROOKLYN,NaN,NaN,NaN,2025,11,November
466264,66651110,2025-10-29 16:00:37,2025-10-31 11:33:00,DCWP,Department of Consumer and Worker Protection,Consumer Complaint,Ticket Seller,Business,NaN,1236,...,"1,006,018",NaN,ONLINE,BRONX,NaN,NaN,NaN,2025,10,October
1174794,65783225,2025-08-07 06:55:29,2025-08-07 10:06:57,DOT,Department of Transportation,Broken Parking Meter,No Receipt,Street,10001.0,WEST 28 STREET,...,"986,393",NaN,PHONE,MANHATTAN,NaN,NaN,NaN,2025,8,August
2166450,64619746,2025-04-11 16:22:49,2025-04-17 11:45:32,DOT,Department of Transportation,Highway Condition,Pothole - Highway,Highway,NaN,NaN,...,"1,000,115",NaN,PHONE,MANHATTAN,NaN,NaN,NaN,2025,4,April
2747820,63777089,2025-01-15 09:13:05,2025-02-11 07:31:48,DOT,Department of Transportation,Broken Parking Meter,No Receipt,Street,10462.0,WESTCHESTER AVENUE,...,"1,023,085",NaN,PHONE,BRONX,NaN,NaN,NaN,2025,1,January
2801240,63704297,2025-01-09 10:34:00,2025-02-19 13:31:04,DOE,Department of Education,School Maintenance,Heating Problem,School,NaN,750,...,"1,030,803",NaN,ONLINE,BRONX,NaN,NaN,NaN,2025,1,January


We are removing all 33 where we do not have location and both Y and X coordinates

In [24]:
# Keep only rows where Location is not null
df = df[df["Location"].notna()]

# Quick check
df["Location"].isna().sum()  # should be 0

0

## We have some missing Compaint Detail, maybe we can replace the name of the Detaio wiht the compaint name

In [25]:
# Filter rows where Complaint Detail is null
missing_detail_rows = df[df["Complaint Detail"].isna()]

# Get unique Complaint values that have missing details
missing_complaints = missing_detail_rows["Complaint"].unique()

# Display them
missing_complaints

array(['Encampment', 'Illegal Fireworks', 'Panhandling',
       'Bike/Roller/Skate', 'Asbestos', 'Urinating in Public',
       'Animal Facility - No Permit', 'Poison Ivy', 'SNW',
       'Bike/Roller/Skate Chronic', 'Squeegee',
       'Homeless Person Assistance', 'ASBESTOS'], dtype=object)

In [26]:
for complaint in missing_complaints:
    rows_with_details = df[(df["Complaint"] == complaint) & (df["Complaint Detail"].notna())]
    print(f"Complaint: {complaint}, Rows with details: {len(rows_with_details)}")
    display(rows_with_details.head()) 

Complaint: Encampment, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Illegal Fireworks, Rows with details: 2


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
3876324,62222993,2024-08-24 22:12:06,2024-08-24 23:50:01,NYPD,New York City Police Department,Illegal Fireworks,Loud Music/Party,Street/Sidewalk,11221.0,41 WEIRFIELD STREET,...,"1,007,803","189,609",ONLINE,BROOKLYN,40.687078,-73.915074,POINT (-73.915073683032 40.687078337762),2024,8,August
3876559,62225406,2024-08-24 21:51:56,2024-08-25 01:54:54,NYPD,New York City Police Department,Illegal Fireworks,Loud Music/Party,Street/Sidewalk,11207.0,824 KNICKERBOCKER AVENUE,...,"1,010,119","191,469",ONLINE,BROOKLYN,40.692177,-73.906716,POINT (-73.906715621362 40.692177140876),2024,8,August


Complaint: Panhandling, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Bike/Roller/Skate, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Asbestos, Rows with details: 1817


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
10,67346601,2025-12-31 16:20:00,2026-01-02 08:45:00,DEP,Department of Environmental Protection,Asbestos,Asbestos Complaint (B1),NaN,11694.0,175 BEACH 101 STREET,...,"1,033,833","151,790",ONLINE,QUEENS,40.583166,-73.821494,POINT (-73.8214939929 40.58316590758),2025,12,December
2909,67350176,2025-12-31 07:26:00,2025-12-31 09:30:00,DEP,Department of Environmental Protection,Asbestos,Asbestos Complaint (B1),NaN,10475.0,2185 REEDS MILL LANE,...,"1,030,751","260,725",PHONE,BRONX,40.882182,-73.831837,POINT (-73.831836729458 40.882181541475),2025,12,December
4230,67336009,2025-12-30 22:29:00,2025-12-31 08:30:00,DEP,Department of Environmental Protection,Asbestos,Asbestos Complaint (B1),NaN,10302.0,1324 FOREST AVENUE,...,"947,074","167,467",PHONE,STATEN ISLAND,40.626257,-74.133925,POINT (-74.1339253738 40.62625673907),2025,12,December
4328,67334705,2025-12-30 22:05:00,2025-12-30 23:00:00,DEP,Department of Environmental Protection,Asbestos,Asbestos Complaint (B1),NaN,10021.0,7 EAST 75 STREET,...,"993,920","221,339",PHONE,MANHATTAN,40.774195,-73.965087,POINT (-73.965086752477 40.774195458987),2025,12,December
4376,67334706,2025-12-30 21:56:00,2026-01-02 08:00:00,DEP,Department of Environmental Protection,Asbestos,Asbestos Complaint (B1),NaN,11218.0,611 DAHILL ROAD,...,"990,194","169,804",PHONE,BROOKLYN,40.632747,-73.978585,POINT (-73.978584851993 40.632747320384),2025,12,December


Complaint: Urinating in Public, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Animal Facility - No Permit, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Poison Ivy, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: SNW, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Bike/Roller/Skate Chronic, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Squeegee, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Complaint: Homeless Person Assistance, Rows with details: 28723


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
216,67347664,2025-12-31 15:36:54,2026-01-01 15:00:01,DHS,Department of Homeless Services,Homeless Person Assistance,Non-Chronic,Park/Playground,11238.0,10 GRAND ARMY PLAZA,...,"992,830","184,409",MOBILE,BROOKLYN,40.672833,-73.969069,POINT (-73.969069245456 40.672832697499),2025,12,December
472,67345295,2025-12-31 14:55:15,2026-01-01 11:41:13,DHS,Department of Homeless Services,Homeless Person Assistance,Non-Chronic,Street/Sidewalk,10011.0,20 WASHINGTON SQUARE NORTH,...,"984,886","205,884",PHONE,MANHATTAN,40.731781,-73.997705,POINT (-73.997705202582 40.731780669245),2025,12,December
483,67353648,2025-12-31 14:53:36,2026-01-01 10:44:09,DHS,Department of Homeless Services,Homeless Person Assistance,Non-Chronic,Store/Commercial,10003.0,104 EAST 14 STREET,...,"987,223","206,751",PHONE,MANHATTAN,40.734160,-73.989273,POINT (-73.989272521475 40.734159894398),2025,12,December
690,67345296,2025-12-31 14:21:46,2026-01-01 14:59:44,DHS,Department of Homeless Services,Homeless Person Assistance,Non-Chronic,Street/Sidewalk,11221.0,570 VAN BUREN STREET,...,"1,004,589","191,388",PHONE,BROOKLYN,40.691969,-73.926657,POINT (-73.926657198789 40.691969250994),2025,12,December
747,67353649,2025-12-31 14:13:47,2025-12-31 16:24:25,DHS,Department of Homeless Services,Homeless Person Assistance,Non-Chronic,Store/Commercial,10025.0,BROADWAY,...,"992,327","229,267",PHONE,MANHATTAN,40.795957,-73.970829,POINT (-73.970828698959 40.795957302803),2025,12,December


Complaint: ASBESTOS, Rows with details: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


so it is almmost always miising detail fot those in a missing_complaints list, except for Asbestos, Illegal Fireworks, and Homeless Person Assistance

In [27]:
df[df["Complaint Detail"]=='Loud Music/Party']

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
21,67349420,2025-12-31 16:17:18,2025-12-31 16:18:36,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,11366.0,78-15 PARSONS BOULEVARD,...,"1,037,000","202,363",ONLINE,QUEENS,40.721959,-73.809697,POINT (-73.809696824262 40.721959131993),2025,12,December
51,67346430,2025-12-31 16:11:36,2026-01-01 05:35:58,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,Street/Sidewalk,11433.0,107-07 157 STREET,...,"1,040,701","192,934",ONLINE,QUEENS,40.696056,-73.796424,POINT (-73.796424037638 40.696055987633),2025,12,December
58,67350589,2025-12-31 16:08:29,2025-12-31 18:01:47,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,11207.0,216 SCHAEFER STREET,...,"1,010,078","190,909",MOBILE,BROOKLYN,40.690640,-73.906866,POINT (-73.906865615486 40.690640190671),2025,12,December
60,67350590,2025-12-31 16:08:09,2025-12-31 19:27:05,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,11213.0,897 EMPIRE BOULEVARD,...,"1,002,926","181,023",MOBILE,BROOKLYN,40.663523,-73.932683,POINT (-73.932682727575 40.663523343151),2025,12,December
78,67351784,2025-12-31 16:03:54,2025-12-31 18:04:56,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,10026.0,1350 5 AVENUE,...,"998,623","230,180",ONLINE,MANHATTAN,40.798455,-73.948088,POINT (-73.948087797467 40.798455241461),2025,12,December
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5597323,59896018,2024-01-01 16:29:06,2024-01-01 16:39:46,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,Street/Sidewalk,10005.0,100 WALL STREET,...,"982,286","196,116",ONLINE,MANHATTAN,40.704970,-74.007084,POINT (-74.007083599796 40.704969661475),2024,1,January
5597337,59894371,2024-01-01 16:27:47,2024-01-01 17:35:27,NYPD,New York City Police Department,Noise - Street/Sidewalk,Loud Music/Party,Street/Sidewalk,11228.0,819 71 STREET,...,"979,985","168,373",MOBILE,BROOKLYN,40.628820,-74.015365,POINT (-74.01536511302 40.628820495338),2024,1,January
5597338,59900327,2024-01-01 16:27:38,2024-01-01 19:39:59,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,10029.0,154 EAST 105 STREET,...,"999,112","227,601",ONLINE,MANHATTAN,40.791376,-73.946327,POINT (-73.946327341838 40.791375777383),2024,1,January
5597349,59898545,2024-01-01 16:25:11,2024-01-01 19:18:54,NYPD,New York City Police Department,Noise - Residential,Loud Music/Party,Residential Building/House,11433.0,107-09 159 STREET,...,"1,041,091","193,174",ONLINE,QUEENS,40.696712,-73.795016,POINT (-73.795015583078 40.696712233714),2024,1,January


In [28]:
df[df["Complaint Detail"].str.contains("Fireworks", case=False, na=False)]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


Some complaints don’t have a Complaint Detail because the complaint itself fully describes the issue. I will replace missing Complaint Detail values with the Complaint value.

In [29]:
df.loc[df["Complaint Detail"].isna(), "Complaint Detail"] = df["Complaint"]

In [30]:
df[df["Complaint Detail"].isna()]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


## we also have null values in Closed Date

- Created Date    0
- Closed Date  136499

In [31]:
missing_closed_rows = df[df["Closed Date"].isna()]

# Get unique Complaint values for those rows
missing_closed_complaints = missing_closed_rows["Complaint"].unique()

# Display them
missing_closed_complaints

array(['For Hire Vehicle Complaint', 'Noise - Helicopter', 'Water System',
       'Graffiti', 'Street Light Condition', 'For Hire Vehicle Report',
       'Sanitation Worker or Vehicle Complaint', 'Illegal Dumping',
       'Indoor Sewage', 'Street Sign - Damaged', 'Damaged Tree',
       'Root/Sewer/Sidewalk Condition', 'Food Poisoning',
       'Street Condition', 'Taxi Complaint', 'Dead/Dying Tree',
       'Dirty Condition', 'Overgrown Tree/Branches', 'Sidewalk Condition',
       'Pet Sale', 'Sewer', 'Maintenance or Facility',
       'Construction Lead Dust', 'Uprooted Stump', 'Lot Condition',
       'Street Sign - Dangling', 'Taxi Report',
       'Tobacco or Non-Tobacco Sale', 'Illegal Tree Damage',
       'Green Taxi Complaint', 'Unleashed Dog', 'Street Sign - Missing',
       'Curb Condition', 'Rodent', 'Derelict Vehicles',
       'Harboring Bees/Wasps', 'Industrial Waste', 'Hazardous Materials',
       'Noise', 'Unsanitary Animal Facility', 'Green Taxi Report',
       'Drinking Wate

I do not think it is relevan to our analysis so I will leave this as is

## Remove agencies that are not neighborhood-quality related

In [32]:
remove_agencies = [
    'TLC',     # taxi / ride complaints
    'DCWP',    # consumer complaints
    'OTI',     # LinkNYC / payphones
]

df = df[~df['Agency'].isin(remove_agencies)]

###  Remove complaint types that are not useful

In [33]:
remove_complaints = [
    'Lost Property',
    'Found Property',
    'Taxi Compliment',
    'Taxi Report',
    'For Hire Vehicle Report',
    'For Hire Vehicle Complaint',
    'Green Taxi Complaint',
    'Green Taxi Report',
    'Dispatched Taxi Complaint',
    'Taxi Licensee Complaint',
    'Public Payphone Complaint',
    'LinkNYC',
    'Consumer Complaint'
]

df = df[~df['Complaint'].isin(remove_complaints)]

### Remove vague / useless rows

In [34]:
df = df[~df['Complaint'].isin(['Unspecified'])]

### Remove information requests (not complaints)

In [35]:
remove_complaint_detail = [
    'Lead Kit Request (Residential) (L10)',
    'New Basket',
    'Replacement Basket',
    'New Bus Stop Shelter Placement'
]

df = df[~df['Complaint Detail'].isin(remove_complaint_detail)]

### Optional: Remove very rare complaint types ????

In [36]:
counts = df['Complaint'].value_counts()

df = df[df['Complaint'].isin(counts[counts > 100].index)]

In [38]:
df['Complaint'].value_counts().head(20)

Complaint
Illegal Parking                  1075646
Noise - Residential               839610
Blocked Driveway                  341836
Noise - Street/Sidewalk           333117
Water System                      141028
Abandoned Vehicle                 136687
Noise - Commercial                131754
Dirty Condition                   119149
Noise                             109902
Street Condition                  101526
Noise - Vehicle                    90888
Encampment                         89405
Traffic Signal Condition           84949
Derelict Vehicles                  84359
Illegal Dumping                    76686
Homeless Person Assistance         76212
Missed Collection                  75897
General Construction/Plumbing      71729
Rodent                             70667
Damaged Tree                       61382
Name: count, dtype: int64

In [39]:
df['Agency'].value_counts()

Agency
NYPD     3243314
DSNY      608630
DOT       357259
DEP       340104
DPR       248014
DOB       204591
DOHMH     172821
DHS       102452
EDC        49086
DOE         3456
OOS         1108
3-1-1          1
Name: count, dtype: int64

the one left for a 311 is noise complaint - so we can leave it

### Create the mapping dictionary

In [40]:
category_map = {

# --------------------
# NOISE
# --------------------
"Noise - Residential": "Noise",
"Noise - Street/Sidewalk": "Noise",
"Noise - Commercial": "Noise",
"Noise - Vehicle": "Noise",
"Noise - Park": "Noise",
"Noise - Helicopter": "Noise",
"Noise": "Noise",

# --------------------
# PARKING & TRAFFIC
# --------------------
"Illegal Parking": "Parking & Traffic",
"Blocked Driveway": "Parking & Traffic",
"Traffic": "Parking & Traffic",

# --------------------
# SANITATION
# --------------------
"Dirty Condition": "Sanitation & Trash",
"Missed Collection": "Sanitation & Trash",
"Illegal Dumping": "Sanitation & Trash",
"Litter Basket Complaint": "Sanitation & Trash",
"Litter Basket Request": "Sanitation & Trash",
"Residential Disposal Complaint": "Sanitation & Trash",
"Commercial Disposal Complaint": "Sanitation & Trash",
"Dumpster Complaint": "Sanitation & Trash",
"Street Sweeping Complaint": "Sanitation & Trash",
"Obstruction": "Sanitation & Trash",
"Dead Animal": "Sanitation & Trash",

# --------------------
# RODENTS
# --------------------
"Rodent": "Rodents & Pests",

# --------------------
# STREET / INFRASTRUCTURE
# --------------------
"Street Condition": "Street & Infrastructure",
"Street Light Condition": "Street & Infrastructure",
"Sidewalk Condition": "Street & Infrastructure",
"Curb Condition": "Street & Infrastructure",
"Traffic Signal Condition": "Street & Infrastructure",
"Broken Parking Meter": "Street & Infrastructure",
"Street Sign - Damaged": "Street & Infrastructure",
"Street Sign - Missing": "Street & Infrastructure",
"Street Sign - Dangling": "Street & Infrastructure",
"Highway Condition": "Street & Infrastructure",
"Bridge Condition": "Street & Infrastructure",

# --------------------
# WATER / SEWER
# --------------------
"Water System": "Water & Sewer",
"Sewer": "Water & Sewer",
"Water Quality": "Water & Sewer",
"Water Conservation": "Water & Sewer",

# --------------------
# BUILDINGS
# --------------------
"Building/Use": "Buildings & Construction",
"General Construction/Plumbing": "Buildings & Construction",
"Construction Safety Enforcement": "Buildings & Construction",
"Elevator": "Buildings & Construction",
"Boilers": "Buildings & Construction",
"Electrical": "Buildings & Construction",
"Plumbing": "Buildings & Construction",
"Facade Insp Safety Pgm": "Buildings & Construction",

# --------------------
# PARKS / TREES
# --------------------
"Damaged Tree": "Parks & Trees",
"Dead/Dying Tree": "Parks & Trees",
"Overgrown Tree/Branches": "Parks & Trees",
"Illegal Tree Damage": "Parks & Trees",
"New Tree Request": "Parks & Trees",
"Root/Sewer/Sidewalk Condition": "Parks & Trees",
"Maintenance or Facility": "Parks & Trees",
"Violation of Park Rules": "Parks & Trees",
"Animal in a Park": "Parks & Trees",

# --------------------
# PUBLIC SAFETY
# --------------------
"Drug Activity": "Public Safety",
"Drinking": "Public Safety",
"Disorderly Youth": "Public Safety",
"Non-Emergency Police Matter": "Public Safety",
"Animal-Abuse": "Public Safety",

# --------------------
# ENVIRONMENT
# --------------------
"Air Quality": "Environment & Air Quality",
"Hazardous Materials": "Environment & Air Quality",
"Asbestos": "Environment & Air Quality"
}

In [41]:
#Apply the mapping
df["Main Complaint Category"] = df["Complaint"].map(category_map)

In [42]:
#Remove rows that did not map
df = df[df["Main Complaint Category"].notna()]

In [43]:
df["Main Complaint Category"].value_counts()

Main Complaint Category
Noise                        1572614
Parking & Traffic            1438023
Sanitation & Trash            385529
Street & Infrastructure       342343
Parks & Trees                 245674
Water & Sewer                 203203
Buildings & Construction      175051
Public Safety                 129778
Rodents & Pests                70667
Environment & Air Quality      25973
Name: count, dtype: int64

In [45]:
#sorted(df["Complaint"].unique())

## Remove additional columns to
- groupby operations faster
- visualizations easier
- code cleaner

In [46]:
df.columns


Index(['Unique Key', 'Created Date', 'Closed Date', 'Agency', 'Agency Name',
       'Complaint', 'Complaint Detail', 'Location Type', 'Incident Zip',
       'Incident Address', 'Street Name', 'Cross Street 1', 'Cross Street 2',
       'Intersection Street 1', 'Intersection Street 2', 'Address Type',
       'City', 'Landmark', 'Facility Type', 'Status', 'Due Date',
       'Resolution Description', 'Resolution Action Updated Date',
       'Community Board', 'BBL', 'Borough', 'X Coordinate', 'Y Coordinate',
       'Open Data Channel Type', 'Park Borough', 'Latitude', 'Longitude',
       'Location', 'Complaint Year', 'Complaint Month', 'Month Name',
       'Main Complaint Category'],
      dtype='object')

**we do not need those:**
BBL
X Coordinate
Y Coordinate
Location Type

Incident Address
Street Name
Cross Street 1
Cross Street 2
Intersection Street 1
Intersection Street 2
Address Type
Landmark
Location

**These describe how agencies processed the complaint: - remove as well**
Due Date
Resolution Description
Resolution Action Updated Date
Facility Type
Open Data Channel Type
Park Borough

for mapping we can keep:
    'Latitude', 'Longitude', 'Borough', 'Community Board'

In [47]:
drop_cols = [
'Incident Address','Street Name','Cross Street 1','Cross Street 2',
'Intersection Street 1','Intersection Street 2','Address Type',
'Landmark','Location',
'Due Date','Resolution Description','Resolution Action Updated Date',
'Facility Type','Open Data Channel Type','Park Borough',
'BBL','X Coordinate','Y Coordinate','Location Type','City'
]

df = df.drop(columns=drop_cols)

In [49]:
df.head()

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Incident Zip,Status,Community Board,Borough,Latitude,Longitude,Complaint Year,Complaint Month,Month Name,Main Complaint Category
0,67347796,2025-12-31 16:23:00,2026-01-01 01:00:00,DEP,Department of Environmental Protection,Water System,Dirty Water (WE),10031.0,Closed,09 MANHATTAN,MANHATTAN,40.823646,-73.954431,2025,12,December,Water & Sewer
2,67347638,2025-12-31 16:22:31,2026-01-02 08:31:35,DOT,Department of Transportation,Street Sign - Missing,Other/Unknown,10314.0,Closed,02 STATEN ISLAND,STATEN ISLAND,40.591542,-74.190961,2025,12,December,Street & Infrastructure
3,67346505,2025-12-31 16:22:09,2026-01-07 20:54:28,DSNY,Department of Sanitation,Dirty Condition,Trash,10026.0,Closed,10 MANHATTAN,MANHATTAN,40.803343,-73.952816,2025,12,December,Sanitation & Trash
4,67351565,2025-12-31 16:21:58,2025-12-31 16:37:20,NYPD,New York City Police Department,Noise - Vehicle,Car/Truck Horn,11366.0,Closed,08 QUEENS,QUEENS,40.721959,-73.809697,2025,12,December,Noise
5,67346439,2025-12-31 16:21:39,2026-01-06 14:53:13,DOT,Department of Transportation,Street Condition,Defective Hardware,11238.0,Closed,02 BROOKLYN,BROOKLYN,40.687820,-73.962694,2025,12,December,Street & Infrastructure


## need to fix zipcode - make it str

In [50]:
# Make sure Incident Zip is numeric (drop decimals) and then convert to string
df['Incident Zip'] = df['Incident Zip'].fillna(0).astype(int).astype(str)

# Optional: pad with leading zeros if needed (NYC zips are 5 digits)
df['Incident Zip'] = df['Incident Zip'].str.zfill(5)

## Lets see what we missing now

In [52]:
df.isnull().sum()

Unique Key                     0
Created Date                   0
Closed Date                70763
Agency                         0
Agency Name                    0
Complaint                      0
Complaint Detail               0
Incident Zip                   0
Status                         0
Community Board                0
Borough                        0
Latitude                       0
Longitude                      0
Complaint Year                 0
Complaint Month                0
Month Name                     0
Main Complaint Category        0
dtype: int64

#### we just converted all zipcodes to str and missing one to 0 so need to chcek how many is still missing

In [55]:
missing_after = (df['Incident Zip'] == '00000').sum()
print("Missing/invalid ZIP codes after cleaning:", missing_after)

Missing/invalid ZIP codes after cleaning: 524


## save your cleaned dataset so we don’t have to repeat all the preprocessing

In [56]:
# Save your cleaned DataFrame to a CSV
df.to_csv("311_cleaned.csv", index=False)

# Optionally, save as a pickle for faster reloading with all dtypes preserved
df.to_pickle("311_cleaned.pkl")

## We have missing zipcodes

- Incident Zip   524 missing zipcodes